05_assessing_code_quality/1_project_structure.md

```diff
@@ -0,0 +1,111 @@
+
+The problem about a folder with a name such as `utils` is that it's too generic and they don't add any semantic meaning to the functions and classes. It's a dumping ground for code that doesn't fit anywhere else. It's okay if you keep them small, but be careful with folders with generic names like `utils`, `functions`, `helper` and `common`: they can quickly become a black hole.
+
+Another problem is the depth of nested folders. The closer a function is to the surface, the easier it is to find. So keep the most important functions and classes close to the surface. That why, in `pandas` you can import the `read_csv` function directly from the `pandas` module. Imagine if you had to import it from `pandas.io.parsers.read_csv`!
+
+### 3. Make things modular
+
+Remember when we said programmers see each new requirement as a new puzzle piece that they must be fit into a puzzle that grow more and more complex?
+
+The solution is to make the puzzle pieces interchangeable. If your puzzle is make only of square pieces, then you can easily swap them around. But if your puzzle is made of pieces of different shapes, then it's considerable harder.
+
+The same applies to your code. If you can swap parts of your code with ease, then your code is modular.
+
+A good strategy is to make your code modular is to think about it as LEGO blocks. Each block should be self-contained and have a single responsibility. Then you have an "entrypoint" (i.e. the main script) that connects the blocks together.
+
+#### What is a "responsibility"?
+
+A responsibility is a reason to change.
+
+As an example, consider a function/module/class that compiles and prints a report. This piece of code can be changed for two reasons:
+
+1. The content of the report could change.
+2. The format of the report could change.
+
+That means that it currently contain two separate responsibilities. Because they are together, if there is a change to the report compilation process, there is a greater danger that the printing code will break. And vice-versa.
+
+## References for project structure
+
+Now that we know what is a good project structure, let's see some references to inspire us. Note how you can easily tell what the project does just by looking at the project structure.
+
+### Cookiecutter Data Science (0h15)
+
+[Link to Cookiecutter Data Science](https://drivendata.github.io/cookiecutter-data-science/)
+
+The Cookiecutter Data Science project is a template for data science projects. It's a good starting point for many projects. It's opinionated, but all its opinions are justified.
+
+After this lesson you should:
+
+- Have a reference on how to layout projects involving data flows.
+
+### Python application layouts (0h15)
+
+![RealPython cover](../images/ba19cadbefac3e8cbea6ab6829321d377e981efc409395804470873bec7e432a.png)
+
+[Link to RealPython guide](https://realpython.com/python-application-layouts/)
```

Comment:
[nitpick] This link requires login

Reply:
I think they only require it after a given number of visits. You can still open in anonymous mode.

The login doesn't require a paid membership, so I think we're ok.